File: archivist/src/main/java/com/zorroa/archivist/domain/Folder.java
package com.zorroa.archivist.domain;

import com.google.common.base.MoreObjects;
import com.zorroa.sdk.search.AssetSearch;

import java.util.Map;
import java.util.Objects;
import java.util.UUID;

public class Folder implements Loggable<UUID> {

    public static final UUID ROOT_ID = UUID.fromString("00000000-0000-0000-0000-000000000000");

    /**
     * Return true if the given folder ID is the root folder's ID.
     * @param id
     * @return
     */
    public static boolean isRoot(UUID id) {
        return ROOT_ID == id;
    }

    /**
     * Return true if the given folder is the root folder.
     * @param folder
     * @return
     */
    public static boolean isRoot(Folder folder) {
        return ROOT_ID == folder.getId();
    }

    private UUID id;
    private UUID parentId;
    private UUID dyhiId;
    private String name;
    private UserBase user;
    private long timeCreated;
    private long timeModified;
    private boolean recursive;
    private boolean dyhiRoot;
    private String dyhiField;
    private int childCount;

    private Acl acl;

    private AssetSearch search;
    private boolean taxonomyRoot;
    private Map<String, Object> attrs;

    public Folder() { }

    public UUID getId() {
        return id;
    }

    public Folder setId(UUID id) {
        this.id = id;
        return this;
    }

    public UUID getParentId() {
        return parentId;
    }

    public Folder setParentId(UUID parentId) {
        this.parentId = parentId;
        return this;
    }

    public String getName() {
        return name;
    }

    public Folder setName(String name) {
        this.name = name;
        return this;
    }

    public UserBase getUser() {
        return user;
    }

    public Folder setUser(UserBase user) {
        this.user = user;
        return this;
    }

    public long getTimeCreated() {
        return timeCreated;
    }

    public Folder setTimeCreated(long timeCreated) {
        this.timeCreated = timeCreated;
        return this;
    }

    public long getTimeModified() {
        return timeModified;
    }

    public Folder setTimeModified(long timeModified) {
        this.timeModified = timeModified;
        return this;
    }

    public boolean isRecursive() {
        return recursive;
    }

    public Folder setRecursive(boolean recursive) {
        this.recursive = recursive;
        return this;
    }

    public AssetSearch getSearch() {
        return search;
    }

    public Folder setSearch(AssetSearch search) {
        this.search = search;
        return this;
    }

    public UUID getDyhiId() {
        return dyhiId;
    }

    public Folder setDyhiId(UUID dyhiId) {
        this.dyhiId = dyhiId;
        return this;
    }

    public boolean isDyhiRoot() {
        return dyhiRoot;
    }

    public Folder setDyhiRoot(boolean dyhiRoot) {
        this.dyhiRoot = dyhiRoot;
        return this;
    }

    public Acl getAcl() {
        return acl;
    }

    public Folder setAcl(Acl acl) {
        this.acl = acl;
        return this;
    }

    public String getDyhiField() {
        return dyhiField;
    }

    public Folder setDyhiField(String dyhiField) {
        this.dyhiField = dyhiField;
        return this;
    }

    public boolean isTaxonomyRoot() {
        return taxonomyRoot;
    }

    public Folder setTaxonomyRoot(boolean taxonomyRoot) {
        this.taxonomyRoot = taxonomyRoot;
        return this;
    }

    public Map<String, Object> getAttrs() {
        return attrs;
    }

    public Folder setAttrs(Map<String, Object> attrs) {
        this.attrs = attrs;
        return this;
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("id", id)
                .add("parentId", parentId)
                .add("name", name)
                .toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Folder that = (Folder) o;
        return Objects.equals(getId(), that.getId());
    }

    public int getChildCount() {
        return childCount;
    }

    public Folder setChildCount(int childCount) {
        this.childCount = childCount;
        return this;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }


    @Override
    public UUID getTargetId() {
        return getId();
    }
}

File: archivist/src/main/kotlin/com/zorroa/archivist/domain/Folder.kt
package com.zorroa.archivist.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import com.google.common.base.MoreObjects
import com.zorroa.archivist.repository.DaoFilter
import com.zorroa.archivist.search.AssetSearch
import io.swagger.annotations.ApiModel
import io.swagger.annotations.ApiModelProperty
import java.util.Objects
import java.util.UUID

class FolderFilter : DaoFilter() {

    var paths: List<String>? = null

    override val sortMap: Map<String, String>?
        get() = mapOf("name" to "str_name")

    override fun build() {
    }
}

fun isRootFolder(folder: Folder): Boolean {
    return folder.name == "/" && folder.parentId == null
}

@ApiModel("Folder", description = "Describes a Folder.")
data class Folder(

    @ApiModelProperty("UUID of the Folder.")
    val id: UUID,

    @ApiModelProperty("Name of the Folder.")
    val name: String,

    @ApiModelProperty("UUID of the this Folder's parent Folder.")
    val parentId: UUID?,

    @ApiModelProperty("UUID of the Organization this Folder belongs to.")
    val organizationId: UUID,

    @ApiModelProperty("UUID of the Dynamic Hierarchy this Folder belongs to.")
    val dyhiId: UUID?,

    @ApiModelProperty("User this Folder belongs to.")
    val user: UserBase,

    @ApiModelProperty("Time this Folder was creates.")
    val timeCreated: Long,

    @ApiModelProperty("Time this Folder was last modified.")
    val timeModified: Long,

    @ApiModelProperty(hidden = true)
    val recursive: Boolean,

    @ApiModelProperty("If true this Folder is the root of a Dynamic Hierarchy.")
    val dyhiRoot: Boolean,

    @ApiModelProperty("Field this Folder is based on in a Dynamic Hierarchy.")
    val dyhiField: String?,

    @ApiModelProperty("Number of children this Folder has.")
    val childCount: Int = 0,

    @ApiModelProperty("ACL applied to this Folder.")
    val acl: Acl? = null,

    @ApiModelProperty("Search filter responsible for populating this Folder.")
    val search: AssetSearch? = null,

    @ApiModelProperty("If true this Folder is the root of a taxonomy.")
    var taxonomyRoot: Boolean = false,

    @ApiModelProperty("Attributes for this Folder.")
    val attrs: Map<String, Any>? = null

) {

    override fun toString(): String {
        return MoreObjects.toStringHelper(this)
                .add("parentId", parentId)
                .add("name", name)
                .toString()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as Folder
        return id == that.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }
}

@ApiModel("Folder Update", description = "Describes an update made to a Folder.")
data class FolderUpdate(

    @ApiModelProperty("Name of the Folder.")
    var name: String,

    @ApiModelProperty("UUID of the Folder's parent Folder.")
    var parentId: UUID?,

    @ApiModelProperty("If true the Folder is recursive.")
    var recursive: Boolean = true,

    @ApiModelProperty("Search filter used to populate the Folder.")
    var search: AssetSearch? = null,

    @ApiModelProperty("Folder's attributes.")
    var attrs: Map<String, Any>? = null

) {

    constructor(folder: Folder) :
            this(folder.name, folder.parentId, folder.recursive, folder.search, folder.attrs)
}

@ApiModel("Folder Spec", description = "Attributes required to create a Folder.")
class FolderSpec {

    @ApiModelProperty("Name of the Folder")
    var name: String? = null

    @ApiModelProperty("UUID of the Folder's parent Folder.")
    var parentId: UUID? = null

    @ApiModelProperty("UUID of a Dynamic Hierarchy to apply to this folder.")
    var dyhiId: UUID? = null

    @ApiModelProperty("Search filter to use to populate this Folder.")
    var search: AssetSearch? = null

    @ApiModelProperty("ACL applied to the Folder.")
    var acl: Acl? = null

    @ApiModelProperty("Folder attributes.")
    var attrs: Map<String, Any>? = null

    @ApiModelProperty("If true the Folder will be recursive.")
    var recursive: Boolean = true

    @JsonIgnore
    var created: Boolean = false

    @JsonIgnore
    var userId: UUID? = null

    constructor()

    constructor(name: String) :
            this(name, null)

    constructor(name: String, folder: Folder) :
            this(name, folder.id)

    constructor(name: String, parentId: UUID?) {
        this.name = name
        this.parentId = parentId
    }

    override fun toString(): String {
        return MoreObjects.toStringHelper(this)
                .add("parentId", parentId)
                .add("name", name)
                .toString()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as FolderSpec
        return parentId == that.parentId && name == that.name
    }

    override fun hashCode(): Int {
        return Objects.hash(parentId, name)
    }
}
